### ISSUE TITLE
Add Origin::as_str method

### ISSUE THREAD

**Issue by sd2k** [SergioBenitez/Rocket#1575]
It would be great to be able to access the entire URI of a route in a non-allocating way.

My current use case is [while creating a suitable label for endpoints](https://github.com/sd2k/rocket_prometheus/blob/6ae16098189033531a69226b9248375528fa4f3d/src/lib.rs#L327) in the [rocket-prometheus](https://github.com/sd2k/rocket_prometheus) crate: currently I call `request.route().unwrap().uri.to_string()` in an `on_response` fairing, which returns something like `"/hello/<name>?<caps>"` (i.e. the URI of the route). It feels like this should be doable without requiring an allocation on every request though, since we already have the URI defined statically.

The `path` and `query` methods already exist to get those two respective elements, but the only way to get the full URI is using the `Display` impl.

I'd be happy to have a go at this and submit a PR but might need a bit of guidance; the `origin.rs` module seems to contain quite a few optimizations!

[Discussed on Matrix here.](https://matrix.to/#/!kDIcCXWSVfdahNCJWq:mozilla.org/$Rfy_1G9Tp979Pkt3LdISApkF-M1vYOtaFT71iWpzcQo?via=matrix.org&via=mozilla.org)

**Comment by SergioBenitez** [comment]
An `Origin` being `: 'static` does _not_ mean that it is composed only of static strings. Indeed, an `Origin` can contain heap-allocated `String`s, and it can contain independent path and query parts, making it impossible to return a single `&str` that represents the entire origin without allocation.

For a `Route`, however, we are able to compute whatever we'd like before the application launches, and we _could_ allocate a single `String` (or a similar proxy) once that captures the entire origin. No changes to `Origin` should be made - in general, such changes would not be feasible without additional allocation for the aforementioned reasons.

It would be nice to clean-up route URIs along the way. I would suggest creating a `RouteUri` type:

```rust
struct RouteUri;

impl RouteUri {
    /// Returns a `RouteUri` with the mount point from `self` removed, if any.
    fn unmounted(&self) -> &RouteUri;

    /// The mount point of this route URI.
    fn mount(&self) -> &str;

    /// The path part of this route URI.
    fn path(&self) -> &str;

    /// The query part of this route URI.
    fn query(&self) -> &str;

    /// The full URI.
    fn as_str(&self) -> &str;
}
```

Then, your application could use: `route.uri().as_str()`. We could implement implement `Deref<Target=str>` to simply make `route.uri()` work as an `&str`.

**Comment by SergioBenitez** [comment]
Closed in 2463637d518f98cf97223e6afde90277af3fcb74.
